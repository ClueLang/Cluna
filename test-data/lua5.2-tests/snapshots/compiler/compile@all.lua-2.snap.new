---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
