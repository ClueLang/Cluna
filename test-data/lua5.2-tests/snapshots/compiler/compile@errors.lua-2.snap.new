---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/errors.lua
---
print("testing errors");
local debug = require("debug")
local mt = getmetatable(_G) || {}
local oldmm = mt.__index
mt.__index = nil
global fn doit(s) {
    local f, msg = load(s)
    if f == nil {
        return msg
    }
    local cond, msg = pcall(f)
    return (!cond) && msg
}
global fn checkmessage(prog, msg) {
    local m = doit(prog)
    assert(string.find(m, msg, 1, true));
}
global fn checksyntax(prog, extra, token, line) {
    local msg = doit(prog)
    if !string.find(token, "^<%a") && !string.find(token, "^char%(") {
        token = "'" .. token .. "'"
    }
    token = string.gsub(token, "(%p)", "%%%1")
    local pt = string.format(`^%%[string ".*"%%]:%d: .- near %s$`, line, token)
    assert(string.find(msg, pt));
    assert(string.find(msg, msg, 1, true));
}
assert(doit("error('hi', 0)") == 'hi');
assert(doit("error()") == nil);
if !_no32 {
    assert(doit("table.unpack({}, 1, n=2^30)"));
}
assert(doit("a=math.sin()"));
assert(!doit("tostring(1)") && doit("tostring()"));
assert(doit("tonumber()"));
assert(doit("repeat until 1; a"));
assert(doit("return;;"));
assert(doit("assert(false)"));
assert(doit("assert(nil)"));
assert(doit("function a (... , ...) end"));
assert(doit("function a (, ...) end"));
assert(doit("local t={}; t = t[#t] + 1"));
checksyntax(`
  local a = {4

`, "'}' expected (to close '{' at line 1)", "<eof>", 3);
checkmessage("a=1; bbbb=2; a=math.sin(3)+bbbb(3)", "global 'bbbb'");
checkmessage("a=1; local a,bbbb=2,3; a = math.sin(1) and bbbb(3)", "local 'bbbb'");
checkmessage("a={}; do local a=1 end a:bbbb(3)", "method 'bbbb'");
checkmessage("local a={}; a.bbbb(3)", "field 'bbbb'");
assert(!string.find(doit("a={13}; local bbbb=1; a[bbbb](3)"), "'bbbb'"));
checkmessage("a={13}; local bbbb=1; a[bbbb](3)", "number");
checkmessage("a=(1)..{}", "a table value");
aaa = nil
checkmessage("aaa.bbb:ddd(9)", "global 'aaa'");
checkmessage("local aaa={bbb=1}; aaa.bbb:ddd(9)", "field 'bbb'");
checkmessage("local aaa={bbb={}}; aaa.bbb:ddd(9)", "method 'ddd'");
checkmessage("local a,b,c; (function () a = b+1 end)()", "upvalue 'b'");
assert(!doit("local aaa={bbb={ddd=next}}; aaa.bbb:ddd(nil)"));
checkmessage("local _ENV = {x={}}; a = a + 1", "global 'a'");
checkmessage("b=1; local aaa='a'; x=aaa+b", "local 'aaa'");
checkmessage("aaa={}; x=3/aaa", "global 'aaa'");
checkmessage("aaa='2'; b=nil;x=aaa*b", "global 'b'");
checkmessage("aaa={}; x=-aaa", "global 'aaa'");
assert(!string.find(doit("aaa={}; x=(aaa or aaa)+(aaa and aaa)"), "'aaa'"));
assert(!string.find(doit("aaa={}; (aaa or aaa)()"), "'aaa'"));
checkmessage("print(print < 10)", "function");
checkmessage("print(print < print)", "two function");
_G.D = debug
checkmessage(`
  -- create light udata
  local x = D.upvalueid(function () return debug end, 1)
  D.setuservalue(x, {})
`, "light userdata");
_G.D = nil
checkmessage("(io.write or print){}", "io.write");
checkmessage("(collectgarbage or print){}", "collectgarbage");
local t = {}
for i = 1, 1000 {
    t[(i)] = "a = x" .. i
}
local s = table.concat(t, "; ")
t = nil
checkmessage(s .. "; a = bbb + 1", "global 'bbb'");
checkmessage("local _ENV=_ENV;" .. s .. "; a = bbb + 1", "global 'bbb'");
checkmessage(s .. "; local t = {}; a = t.bbb + 1", "field 'bbb'");
checkmessage(s .. "; local t = {}; t:bbb()", "method 'bbb'");
checkmessage(`aaa=9
repeat until 3==3
local x=math.sin(math.cos(3))
if math.sin(1) == x then return math.sin(1) end   -- tail call
local a,b = 1, {
  {x='a'..'b'..'c', y='b', z=x},
  {1,2,3,4,5} or 3+3<=3+3,
  3+1>3+1,
  {d = x and aaa[x or y]}}
`, "global 'aaa'");
checkmessage(`
local x,y = {},1
if math.sin(1) == 0 then return 3 end    -- return
x.a()`, "field 'a'");
checkmessage(`
prefix = nil
insert = nil
while 1 do
  local a
  if nil then break end
  insert(prefix, a)
end`, "global 'insert'");
checkmessage(`  -- tail call
  return math.sin("a")
`, "'sin'");
checkmessage(`collectgarbage("nooption")`, "invalid option");
checkmessage(`x = print .. "a"`, "concatenate");
checkmessage("getmetatable(io.stdin).__gc()", "no value");
checkmessage(`
local Var
local function main()
  NoSuchName (function() Var=0 end)
end
main()
`, "global 'NoSuchName'");
print('+');
a = {}
setmetatable(a, {
    __index = string
});
checkmessage("a:sub()", "bad self");
checkmessage("string.sub('a', {})", "#2");
checkmessage("('a'):sub{}", "#1");
checkmessage("table.sort({1,2,3}, table.sort)", "'table.sort'");
checkmessage("string.gsub('s', 's', setmetatable)", "setmetatable'");
global fn f(n) {
    local c = coroutine.create(f)
    local a, b = coroutine.resume(c)
    return b
}
assert(string.find(f(), "C stack overflow"));
checkmessage("coroutine.yield()", "outside a coroutine");
f1 = fn () {
    table.sort({
        1, 
        2, 
        3
    }, coroutine.yield);
}
f = coroutine.wrap(fn () {
    return pcall(f1)
})
assert(string.find(select(2, f()), "yield across"));
idsize = 60 - 1
local fn checksize(source) {
    local _, msg = load("x", source)
    msg = string.match(msg, "^([^:]*):")
    assert(msg::len() <= idsize);
}
for i = 60 - 10, 60 + 10 {
    checksize("@" .. string.rep("x", i));
    checksize(string.rep("x", i - 10));
    checksize("=" .. string.rep("x", i));
}
local fn lineerror(s, l) {
    local err, msg = pcall(load(s))
    local line = string.match(msg, ":(%d+):")
    assert((line && line + 0) == l);
}
lineerror("local a\n for i=1,'a' do \n print(i) \n end", 2);
lineerror("\n local a \n for k,v in 3 \n do \n print(k) \n end", 3);
lineerror("\n\n for k,v in \n 3 \n do \n print(k) \n end", 4);
lineerror("function a.x.y ()\na=a+1\nend", 1);
lineerror("a = \na\n+\n{}", 3);
lineerror("a = \n3\n+\n(\n4\n/\nprint)", 6);
lineerror("a = \nprint\n+\n(\n4\n/\n7)", 3);
lineerror("a\n=\n-\n\nprint\n;", 3);
lineerror(`
a
(
23)
`, 1);
lineerror(`
local a = {x = 13}
a
.
x
(
23
)
`, 2);
lineerror(`
local a = {x = 13}
a
.
x
(
23 + a
)
`, 6);
local p = `
function g() f() end
function f(x) error('a', X) end
g()
`
X = 3
lineerror((p), 3);
X = 0
lineerror((p), nil);
X = 1
lineerror((p), 2);
X = 2
lineerror((p), 1);
if !_soft {
    C = 0
    local l = debug.getinfo(1, "l").currentline
    global fn y() {
        C = C + 1
        y();
    }
    local fn checkstackmessage(m) {
        return (string.find(m, "^.-:%d+: stack overflow"))
    }
    assert(checkstackmessage(doit('y()')));
    print('+');
    assert(checkstackmessage(doit('y()')));
    print('+');
    assert(checkstackmessage(doit('y()')));
    print('+');
    C = 0
    local l1
    local fn g(x) {
        l1 = debug.getinfo(x, "l").currentline
        y();
    }
    local _, stackmsg = xpcall(g, debug.traceback, 1)
    print('+');
    local stack = {}
    for line with string.gmatch(stackmsg, "[^\n]*") {
        local curr = string.match(line, ":(%d+):")
        if curr {
            table.insert(stack, tonumber(curr));
        }
    }
    local i = 1
    while stack[(i)] ~= l1 {
        assert(stack[(i)] == l);
        i = i + 1
    }
    assert(i > 15);
    local res, msg = xpcall(error, error)
    assert(!res && type(msg) == 'string');
    print('+');
    local fn f(x) {
        if x == 0 {
            error('a\n');
        } else {
            local aux = fn () {
                return f(x - 1)
            }
            local a, b = xpcall(aux, aux)
            return a, b
        }
    }
    f(3);
    local fn loop(x, y, z) {
        return 1 + loop(x, y, z)
    }
    local res, msg = xpcall(loop, fn (m) {
        assert(string.find(m, "stack overflow"));
        local res, msg = pcall(loop)
        assert(string.find(msg, "error handling"));
        assert(math.sin(0) == 0);
        return 15
    })
    assert(msg == 15);
    res, msg = pcall(fn () {
        for i = 999900, 1000000, 1 {
            table.unpack({}, 1, i);
        }
    })
    assert(string.find(msg, "too many results"));
}
global fn f() {
    error({
        msg = 'x'
    });
}
res, msg = xpcall(f, fn (r) {
    return {
        msg = r.msg .. 'y'
    }
})
assert(msg.msg == 'xy');
a, b, c = xpcall(string.find, error, "alo", "al")
assert(a && b == 1 && c == 2);
a, b, c = xpcall(string.find, fn (x) {
    return {}
}, true, "al")
assert(!a && type(b) == "table" && c == nil);
print('+');
checksyntax("syntax error", "", "error", 1);
checksyntax("1.000", "", "1.000", 1);
checksyntax("[[a]]", "", "[[a]]", 1);
checksyntax("'aa'", "", "'aa'", 1);
checksyntax("\255a = 1", "", "char(255)", 1);
doit('I = load("a=9+"); a=3');
assert(a == 3 && I == nil);
print('+');
lim = 1000
if _soft {
    lim = 100
}
for i = 1, lim {
    doit('a = ');
    doit('a = 4+nil');
}
local fn testrep(init, rep) {
    local s = "local a; " .. init .. string.rep(rep, 400)
    local a, b = load(s)
    assert(!a && string.find(b, "levels"));
}
testrep("a=", "{");
testrep("a=", "(");
testrep("", "a(");
testrep("", "do ");
testrep("", "while a do ");
testrep("", "if a then else ");
testrep("", "function foo () ");
testrep("a=", "a..");
testrep("a=", "a^");
local s = ("a,")::rep(200) .. "a=nil"
local a, b = load(s)
assert(!a && string.find(b, "levels"));
local lim = 127
local s = "local function fooA ()\n  local "
for j = 1, lim {
    s = s .. "a" .. j .. ", "
}
s = s .. "b,c\n"
s = s .. "local function fooB ()\n  local "
for j = 1, lim {
    s = s .. "b" .. j .. ", "
}
s = s .. "b\n"
s = s .. "function fooC () return b+c"
local c = 1 + 2
for j = 1, lim {
    s = s .. "+a" .. j .. "+b" .. j
    c = c + 2
}
s = s .. "\nend  end end"
local a, b = load(s)
assert(c > 255 && string.find(b, "too many upvalues") && string.find(b, "line 5"));
s = "\nfunction foo ()\n  local "
for j = 1, 300 {
    s = s .. "a" .. j .. ", "
}
s = s .. "b\n"
local a, b = load(s)
assert(string.find(b, "line 2"));
mt.__index = oldmm
print('OK');
