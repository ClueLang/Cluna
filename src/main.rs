#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
