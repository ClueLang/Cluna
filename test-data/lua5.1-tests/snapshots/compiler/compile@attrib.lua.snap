---
source: src/main.rs
expression: compiled
input_file: test-data/lua5.1-tests/attrib.lua
---
{
    print("testing require");
    assert(require("string") == string);
    assert(require("math") == math);
    assert(require("table") == table);
    assert(require("io") == io);
    assert(require("os") == os);
    assert(require("debug") == debug);
    assert(require("coroutine") == coroutine);
    assert(type(package.path) == "string");
    assert(type(package.cpath) == "string");
    assert(type(package.loaded) == "table");
    assert(type(package.preload) == "table");
    local DIR = "libs/"
    local fn createfiles(files, preextras, posextras) {
        for n, c with pairs(files) {
            io.output(DIR .. n);
            io.write(string.format(preextras, n));
            io.write(c);
            io.write(string.format(posextras, n));
            io.close(io.output());
        }
    }
    global fn removefiles(files) {
        for n with pairs(files) {
            os.remove(DIR .. n);
        }
    }
    local files = {
        "A.lua" = "", 
        "B.lua" = "assert(...=='B');require 'A'", 
        "A.lc" = "", 
        "A" = "", 
        "L" = "", 
        "XXxX" = "", 
        "C.lua" = "package.loaded[...] = 25; require'C'"
    }
    AA = nil
    local extras = `
NAME = '%s'
REQUIRED = ...
return AA`
    createfiles(files, "", extras);
    local oldpath = package.path
    package.path = string.gsub("D/?.lua;D/?.lc;D/?;D/??x?;D/L", "D/", DIR)
    local try = fn (p, n, r) {
        NAME = nil
        local rr = require(p)
        assert(NAME == n);
        assert(REQUIRED == p);
        assert(rr == r);
    }
    assert(require("C") == 25);
    assert(require("C") == 25);
    AA = nil
    try('B', 'B.lua', true);
    assert(package.loaded.B);
    assert(require("B") == true);
    assert(package.loaded.A);
    package.loaded.A = nil
    try('B', nil, true);
    try('A', 'A.lua', true);
    package.loaded.A = nil
    os.remove(DIR .. 'A.lua');
    AA = {}
    try('A', 'A.lc', AA);
    assert(require("A") == AA);
    AA = false
    try('K', 'L', false);
    try('K', 'L', false);
    assert(rawget(_G, "_REQUIREDNAME") == nil);
    AA = "x"
    try("X", "XXxX", AA);
    removefiles(files);
    package.path = string.gsub("D/?.lua;D/?/init.lua", "D/", DIR)
    files = {
        "P1/init.lua" = "AA = 10", 
        "P1/xuxu.lua" = "AA = 20"
    }
    createfiles(files, "module(..., package.seeall)\n", "");
    AA = 0
    local m = assert(require("P1"))
    assert(m == P1 && m._NAME == "P1" && AA == 0 && m.AA == 10);
    assert(require("P1") == P1 && P1 == m);
    assert(require("P1") == P1);
    assert(P1._PACKAGE == "");
    local m = assert(require("P1.xuxu"))
    assert(m == P1.xuxu && m._NAME == "P1.xuxu" && AA == 0 && m.AA == 20);
    assert(require("P1.xuxu") == P1.xuxu && P1.xuxu == m);
    assert(require("P1.xuxu") == P1.xuxu);
    assert(require("P1") == P1);
    assert(P1.xuxu._PACKAGE == "P1.");
    assert(P1.AA == 10 && P1._PACKAGE == "");
    assert(P1._G == _G && P1.xuxu._G == _G);
    removefiles(files);
    package.path = ""
    assert(!pcall(require, "file_does_not_exist"));
    package.path = "??\0?"
    assert(!pcall(require, "file_does_not_exist1"));
    package.path = oldpath
    local fname = "file_does_not_exist2"
    local m, err = pcall(require, fname)
    for t with string.gmatch(package.path .. ";" .. package.cpath, "[^;]+") {
        t = string.gsub(t, "?", fname)
        assert(string.find(err, t, 1, true));
    }
    local fn import(...) {
        local f = {
            ...
        }
        return fn (m) {
            for i = 1, #f {
                m[(f[(i)])] = _G[(f[(i)])]
            }
        }
    }
    local assert, module, package = assert, module, package
    X = nil
    x = 0
    assert(_G.x == 0);
    module("X");
    x = 1
    assert(_M.x == 1);
    module("X.a.b.c");
    x = 2
    assert(_M.x == 2);
    module("X.a.b", package.seeall);
    x = 3
    assert(X._NAME == "X" && X.a.b.c._NAME == "X.a.b.c" && X.a.b._NAME == "X.a.b");
    assert(X._M == X && X.a.b.c._M == X.a.b.c && X.a.b._M == X.a.b);
    assert(X.x == 1 && X.a.b.c.x == 2 && X.a.b.x == 3);
    assert(X._PACKAGE == "" && X.a.b.c._PACKAGE == "X.a.b." && X.a.b._PACKAGE == "X.a.");
    assert(_PACKAGE .. "c" == "X.a.c");
    assert(X.a._NAME == nil && X.a._M == nil);
    module("X.a", import("X"));
    x = 4
    assert(X.a._NAME == "X.a" && X.a.x == 4 && X.a._M == X.a);
    module("X.a.b", package.seeall);
    assert(x == 3);
    x = 5
    assert(_NAME == "X.a.b" && X.a.b.x == 5);
    assert(X._G == nil && X.a._G == nil && X.a.b._G == _G && X.a.b.c._G == nil);
    setfenv(1, _G);
    assert(x == 0);
    assert(!pcall(module, "x"));
    assert(!pcall(module, "math.sin"));
    local p = ""
    local f, err, when = package.loadlib("libs/lib1.so", p .. "luaopen_lib1")
    if !f {
        (Message || print)('\a\n >>> cannot load dynamic library <<<\n\a');
        print(err, when);
    } else {
        f();
        assert(require("lib1") == lib1);
        collectgarbage();
        assert(lib1.id("x") == "x");
        f = assert(package.loadlib("libs/lib1.so", p .. "anotherfunc"))
        assert(f(10, 20) == "1020\n");
        f, err, when = package.loadlib("libs/lib1.so", p .. "xuxu")
        assert(!f && type(err) == "string" && when == "init");
        package.cpath = "libs/?.so"
        require("lib2");
        assert(lib2.id("x") == "x");
        local fs = require("lib1.sub")
        assert(fs == lib1.sub && next(lib1.sub) == nil);
        module("lib2", package.seeall);
        f = require("-lib2")
        assert(f.id("x") == "x" && _M == f && _NAME == "lib2");
        module("lib1.sub", package.seeall);
        assert(_M == fs);
        setfenv(1, _G);
    }
    f, err, when = package.loadlib("donotexist", p .. "xuxu")
    assert(!f && type(err) == "string" && (when == "open" || when == "absent"));
    {
        local p = package
        package = {}
        p.preload.pl = fn (...) {
            module(...);
            global fn xuxu(x) {
                return x + 20
            }
        }
        require("pl");
        assert(require("pl") == pl);
        assert(pl.xuxu(10) == 30);
        package = p
        assert(type(package.path) == "string");
    }
}
print('+');
print("testing assignments, logical operators, and constructors");
local res, res2 = 27
a, b = 1, 2 + 3
assert(a == 1 && b == 5);
a = {}
global fn f() {
    return 10, 11, 12
}
a.x, b, a[(1)] = 1, 2, f()
assert(a.x == 1 && b == 2 && a[(1)] == 10);
a[(f())], b, a[(f() + 3)] = f(), a, 'x'
assert(a[(10)] == 10 && b == a && a[(13)] == 'x');
{
    local f = fn (n) {
        local x = {}
        for i = 1, n {
            x[(i)] = i
        }
        return unpack(x)
    }
    local a, b, c
    a, b = 0, f(1)
    assert(a == 0 && b == 1);
    A, b = 0, f(1)
    assert(A == 0 && b == 1);
    a, b, c = 0, 5, f(4)
    assert(a == 0 && b == 5 && c == 1);
    a, b, c = 0, 5, f(0)
    assert(a == 0 && b == 5 && c == nil);
}
a, b, c, d = 1 && nil, 1 || nil, (1 && (nil || 1)), 6
assert(!a && b && c && d == 6);
d = 20
a, b, c, d = f()
assert(a == 10 && b == 11 && c == 12 && d == nil);
a, b = f(), 1, 2, 3, f()
assert(a == 10 && b == 1);
assert(a < b == false && a > b == true);
assert((10 && 2) == 2);
assert((10 || 2) == 10);
assert((10 || assert(nil)) == 10);
assert(!(nil && assert(nil)));
assert((nil || "alo") == "alo");
assert((nil && 10) == nil);
assert((false && 10) == false);
assert((true || 10) == true);
assert((false || 10) == 10);
assert(false ~= nil);
assert(nil ~= false);
assert(!nil == true);
assert(!!nil == false);
assert(!!1 == true);
assert(!!a == true);
assert(!!(6 || nil) == true);
assert(!!(nil && 56) == false);
assert(!!(nil && true) == false);
print('+');
a = {}
a[(true)] = 20
a[(false)] = 10
assert(a[(1 < 2)] == 20 && a[(1 > 2)] == 10);
global fn f(a) {
    return a
}
local a = {}
for i = 3000,  - 3000,  - 1 {
    a[(i)] = i
}
a[(10e30)] = "alo"
a[(true)] = 10
a[(false)] = 20
assert(a[(10e30)] == 'alo' && a[(!1)] == 20 && a[(10 < 20)] == 10);
for i = 3000,  - 3000,  - 1 {
    assert(a[(i)] == i);
}
a[(print)] = assert
a[(f)] = print
a[(a)] = a
assert(a[(a)][(a)][(a)][(a)][(print)] == assert);
a[(print)](a[(a[(f)])] == a[(print)]);
a = nil
a = {
    10, 
    9, 
    8, 
    7, 
    6, 
    5, 
    4, 
    3, 
    2, 
     - 3 = 'a', 
    f = print, 
    a = 'a', 
    b = 'ab'
}
a, a.x, a.y = a, a[( - 3)]
assert(a[(1)] == 10 && a[( - 3)] == a.a && a[(f)] == print && a.x == 'a' && !a.y);
a[(1)], f(a)[(2)], b, c = {
    'alo' = assert
}, 10, a[(1)], a[(f)], 6, 10, 23, f(a), 2
a[(1)].alo(a[(2)] == 10 && b == 10 && c == print);
a[(2 ^ 31)] = 10
a[(2 ^ 31 + 1)] = 11
a[( - 2 ^ 31)] = 12
a[(2 ^ 32)] = 13
a[( - 2 ^ 32)] = 14
a[(2 ^ 32 + 1)] = 15
a[(10 ^ 33)] = 16
assert(a[(2 ^ 31)] == 10 && a[(2 ^ 31 + 1)] == 11 && a[( - 2 ^ 31)] == 12 && a[(2 ^ 32)] == 13 && a[( - 2 ^ 32)] == 14 && a[(2 ^ 32 + 1)] == 15 && a[(10 ^ 33)] == 16);
a = nil
{
    local a, i, j, b
    a = {
        'a', 
        'b'
    }
    i = 1
    j = 2
    b = a
    i, a[(i)], a, j, a[(j)], a[(i + j)] = j, i, i, b, j, i
    assert(i == 2 && b[(1)] == 1 && a == 1 && j == b && b[(2)] == 2 && b[(3)] == 1);
}
print('OK');
return res
